internal/k8s/controllers/http_route_controller.go

```diff
@@ -64,9 +65,69 @@ func (r *HTTPRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			&source.Kind{Type: &gateway.ReferencePolicy{}},
 			handler.EnqueueRequestsFromMapFunc(r.referencePolicyToRouteRequests),
 		).
+		Watches(
+			&source.Kind{Type: &corev1.Service{}},
+			handler.EnqueueRequestsFromMapFunc(r.serviceToRouteRequests),
+		).
 		Complete(gatewayclient.NewRequeueingMiddleware(r.Log, r))
 }
 
+// serviceToRouteRequests builds a list of HTTPRoutes that need to be reconciled
+// based on changes to a Service
+func (r *HTTPRouteReconciler) serviceToRouteRequests(object client.Object) []reconcile.Request {
+	service := object.(*corev1.Service)
+
+	routes := r.getRoutesAffectedByService(service)
+	var requests []reconcile.Request
+
+	for _, route := range routes {
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      route.Name,
+				Namespace: route.Namespace,
+			},
+		})
+	}
+
+	return requests
+}
+
+// getRoutesAffectedByService retrieves all HTTPRoutes potentially impacted
+// by the Service being modified. This is done by filtering to HTTPRoutes that
+// have a backendRef matching the Service's namespace and name.
+func (r *HTTPRouteReconciler) getRoutesAffectedByService(service *corev1.Service) []gateway.HTTPRoute {
+	var matches []gateway.HTTPRoute
+
+	routes, err := r.Client.GetHTTPRoutes(r.Context)
+	if err != nil {
+		r.Log.Error("error fetching routes", err)
+		return matches
+	}
+
+	// Return any routes that have a backend reference to service
+	for _, route := range routes {
+	nextRoute:
+		for _, rule := range route.Spec.Rules {
+			for _, ref := range rule.BackendRefs {
+				// The BackendRef may or may not specify a namespace, defaults to route's namespace
+				refNamespace := route.Namespace
+				if ref.Namespace != nil && *ref.Namespace != "" {
+					refNamespace = string(*ref.Namespace)
+				}
```

Comment:
This feels a little awkward (and I'm not sure if the redundant empty check is necessary) with the mixed type comparison below (namespace string, name ObjectName), but the logic seems okay, so no strong preference on any changes.

Reply:
Removed redundant empty check. Not really a way around the mixed type comparisons 😕 . IME the large amount of casting required is the big downside of adding custom types for fields instead of just using primitives.